src/helper/convertColumnElementsToData.js

```diff
@@ -58,47 +100,53 @@ function convertColumnElementsToData(childComponents) {
     children.push(child);
   });
 
-  const elementTemplates = {
-    cell: [],
-    footer: [],
-    groupHeader: [],
-    header: [],
-  };
-
-  const columnProps = [];
-  const hasGroupHeader =
+  const columnElements = getElementsContainer();
+  const columnGroupElements = getElementsContainer();
+  const elementTemplates = getElementTemplates();
+  const useGroupHeader =
     children.length && children[0].type.__TableColumnGroup__;
-  if (hasGroupHeader) {
-    const columnGroupProps = map(children, _extractProps);
-    forEach(children, (columnGroupElement, index) => {
-      elementTemplates.groupHeader.push(columnGroupElement.props.header);
-
-      React.Children.forEach(columnGroupElement.props.children, (child) => {
-        const column = _extractProps(child);
-        column.groupIdx = index;
-        columnProps.push(column);
-        _extractTemplates(elementTemplates, child);
-      });
-    });
-
-    return {
-      columnGroupProps,
-      columnProps,
-      elementTemplates,
-      useGroupHeader: true,
-    };
+
+  let columnIndex = 0;
+  let columnGroupIndex = 0;
+
+  if (useGroupHeader) {
+    for (const columnGroupReactElement of children) {
+      const cellGroupType = getCellGroupType(columnGroupReactElement);
+      const columnGroupProps = _extractProps(columnGroupReactElement);
+      columnGroupProps.index = columnGroupIndex;
+      columnGroupElements[cellGroupType].push(columnGroupProps);
+      elementTemplates.groupHeader.push(columnGroupReactElement.props.header);
+
+      React.Children.forEach(
+        columnGroupReactElement.props.children,
+        (columnReactElement) => {
+          const columnProps = _extractProps(columnReactElement);
+          columnProps.index = columnIndex;
+          columnProps.groupIdx = columnGroupIndex;
+          columnElements[cellGroupType].push(columnProps);
+          _extractTemplates(elementTemplates, columnReactElement);
+          columnIndex++;
+        }
+      );
+
+      columnGroupIndex++;
+    }
+  } else {
+    for (const columnReactElement of children) {
+      const cellGroupType = getCellGroupType(columnReactElement);
+      const columnProps = _extractProps(columnReactElement);
+      columnProps.index = columnIndex;
+      columnElements[cellGroupType].push(columnProps);
+      _extractTemplates(elementTemplates, columnReactElement);
+      columnIndex++;
+    }
   }
 
-  // Use a default column group
-  forEach(children, (child) => {
-    columnProps.push(_extractProps(child));
-    _extractTemplates(elementTemplates, child);
-  });
   return {
-    columnGroupProps: [],
-    columnProps,
+    columnGroupElements,
+    columnElements,
```

Comment:
`columnGroupElements` groups column props by `cellGroupType`, where as `columnProps` is a flat ordered list of `column prop` objects.